### PR TITLE
QF-1514 Disable opting out of all standard versions

### PR DESCRIPTION
**Pull request by SawNawDfE**

### CODE REVIEW

File: src/SFA.DAS.AssessorService.Web/Controllers/Apply/StandardController.cs [hunk 1]
```diff
@@ -568,6 +576,7 @@ public async Task<IActionResult> OptOutStandardVersion(OptOutStandardVersionView
 
         [HttpGet("standard/opt-out/{referenceNumber}/{version}/confirmation", Name = OptOutStandardVersionConfirmationRouteGet)]
         [PrivilegeAuthorize(Privileges.ApplyForStandard)]
+        [ModelStatePersist(ModelStatePersist.RestoreEntry)]
```
Comment (ChrisJWoodcock) [inline src/SFA.DAS.AssessorService.Web/Controllers/Apply/StandardController.cs:579]: Why is this here?

File: src/SFA.DAS.AssessorService.Web/Controllers/Apply/StandardController.cs [hunk 2]
```diff
@@ -580,6 +589,11 @@ public async Task<IActionResult> OptOutStandardVersionConfirmation(string refere
             if (!standardVersions?.Any() ?? false)
                 throw new NotFoundException($"The standard reference {referenceNumber} cannot be found");
 
+            if (!ModelState.IsValid)
```
Comment (ChrisJWoodcock) [inline src/SFA.DAS.AssessorService.Web/Controllers/Apply/StandardController.cs:592]: This is going round in a circle - the model would not be invalid here anyway?

File: src/SFA.DAS.AssessorService.Web/ViewModels/Standard/OptOutStandardVersionViewModel.cs [hunk 3]
```diff
@@ -11,5 +15,7 @@ public class OptOutStandardVersionViewModel
 
         public DateTime? EffectiveFrom { get; set; }
         public DateTime? EffectiveTo { get; set; }
+ 
+        public IEnumerable<StandardVersion> ApprovedVersions { get; set; }
```
Comment (ChrisJWoodcock) [inline src/SFA.DAS.AssessorService.Web/ViewModels/Standard/OptOutStandardVersionViewModel.cs:19]: Is this used?

File: src/SFA.DAS.AssessorService.Web/Views/Standard/OptOutStandardVersion.cshtml [hunk 4]
```diff
@@ -14,18 +14,19 @@
     <div class="govuk-grid-row">
         <div class="govuk-grid-column-two-thirds">
 
+            @if (!ViewData.ModelState.IsValid)
+            {
+                <partial name="~/Views/Standard/_OptOutValidationErrorPartial.cshtml" model=@Model />
```
Comment (ChrisJWoodcock) [inline src/SFA.DAS.AssessorService.Web/Views/Standard/OptOutStandardVersion.cshtml:19]: partials are great if you are using them multiple times, but they make the code less readable - so avoid them if it's a one off

File: src/SFA.DAS.AssessorService.Web/Views/Standard/OptOutStandardVersion.cshtml [hunk 5]
```diff
@@ -58,6 +59,8 @@
                 <input asp-for="Version" value="@Model.Version" type="hidden" />
                 <input asp-for="EffectiveFrom" value="@Model.EffectiveFrom" type="hidden" />
                 <input asp-for="EffectiveTo" value="@Model.EffectiveTo" type="hidden" />
+                <input asp-for="ApprovedVersions" value="@Model.ApprovedVersions" type="hidden" />
```
Comment (ChrisJWoodcock) [inline src/SFA.DAS.AssessorService.Web/Views/Standard/OptOutStandardVersion.cshtml:62]: What is this for - and isn't this an enumerable anyway - you can't have a hidden for that

**Review by ChrisJWoodcock — CHANGES_REQUESTED** [review]

File: src/SFA.DAS.AssessorService.Web.UnitTests/ViewModelTests/StandardDetailsViewModelTests/WhenCallingCannotOptOut.cs [hunk 6]
```diff
@@ -0,0 +1,68 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using NUnit.Framework;
+using SFA.DAS.AssessorService.Api.Types.Models.Standards;
+using SFA.DAS.AssessorService.Web.ViewModels.Standard;
+using SFA.DAS.Testing.AutoFixture;
+
+namespace SFA.DAS.AssessorService.Web.UnitTests.ViewModelTests.StandardDetailsViewModelTests;
```
[the remark this answers is not in the document]
Comment (SawNawDfE) [reply src/SFA.DAS.AssessorService.Web.UnitTests/ViewModelTests/StandardDetailsViewModelTests/WhenCallingCannotOptOut.cs]: Changed

File: src/SFA.DAS.AssessorService.Web.UnitTests/ViewModelTests/StandardDetailsViewModelTests/WhenCallingCannotOptOut.cs [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+
+public class WhenCallingCannotOptOut
+{
+    [Test]
+    public void And_EffectiveToDateIsNull_ReturnsTrue()
+    {
+        var version = new StandardVersion() { EffectiveTo = null };
+        var sut = new StandardDetailsViewModel() { ApprovedVersions = new() { version } };
```
[the remark this answers is not in the document]
Comment (SawNawDfE) [reply src/SFA.DAS.AssessorService.Web.UnitTests/ViewModelTests/StandardDetailsViewModelTests/WhenCallingCannotOptOut.cs]: Removed

File: src/SFA.DAS.AssessorService.Web/Controllers/Apply/StandardController.cs (hunk 1 above)
Comment (SawNawDfE) [reply src/SFA.DAS.AssessorService.Web/Controllers/Apply/StandardController.cs]: Removed

File: src/SFA.DAS.AssessorService.Web/Controllers/Apply/StandardController.cs (hunk 2 above)
Comment (SawNawDfE) [reply src/SFA.DAS.AssessorService.Web/Controllers/Apply/StandardController.cs]: Removed

File: src/SFA.DAS.AssessorService.Web/Views/Standard/OptOutStandardVersion.cshtml (hunk 4 above)
Comment (SawNawDfE) [reply src/SFA.DAS.AssessorService.Web/Views/Standard/OptOutStandardVersion.cshtml]: Removed

File: src/SFA.DAS.AssessorService.Web/Views/Standard/OptOutStandardVersion.cshtml (hunk 5 above)
Comment (SawNawDfE) [reply src/SFA.DAS.AssessorService.Web/Views/Standard/OptOutStandardVersion.cshtml]: Removed

File: src/SFA.DAS.AssessorService.Web.UnitTests/ViewModelTests/StandardDetailsViewModelTests/WhenCallingCannotOptOut.cs [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+        {
+            var sut = new StandardDetailsViewModel() { ApprovedVersions = new List<StandardVersion>() { version } };
+
+            Assert.That(sut.CannotOptOut(version), Is.True);
+        }
+
+        [Test, RecursiveMoqAutoData]
+        public void And_MultipleApprovedVersionsExist_ReturnsFalse(StandardVersion versionOne, StandardVersion versionTwo)
+        {
+            var sut = new StandardDetailsViewModel() { ApprovedVersions = new() { versionOne, versionTwo } };
```
Comment (ChrisJWoodcock) [inline src/SFA.DAS.AssessorService.Web.UnitTests/ViewModelTests/StandardDetailsViewModelTests/WhenCallingCannotOptOut.cs:22]: Convert this new() to its Type() form

**Review by ChrisJWoodcock — CHANGES_REQUESTED** [review]

File: src/SFA.DAS.AssessorService.Web.UnitTests/ViewModelTests/StandardDetailsViewModelTests/WhenCallingCannotOptOut.cs (hunk 8 above)
Comment (SawNawDfE) [reply src/SFA.DAS.AssessorService.Web.UnitTests/ViewModelTests/StandardDetailsViewModelTests/WhenCallingCannotOptOut.cs]: Changed them all

**Review by ChrisJWoodcock — APPROVED** [review]

File: src/SFA.DAS.AssessorService.Web/Views/Standard/_OptOutInsetPartial.cshtml [hunk 9]
```diff
@@ -0,0 +1,15 @@
+﻿<div class="govuk-inset-text" wfd-id="10">
+    <strong>Opting in and out of a standard version</strong>
+
+    <p>You can opt in to a standard version if you want to record outcomes against it.</p>
+
+    <p>You can opt out of any standard version that you no longer want to record outcomes for. You can opt in at any time.</p>
+
+    <p>To record outcomes for a standard version, you must be opted in to at least one standard version.</p>
+
+    <p>You are required to have at least one approved standard version. If you want to opt out of all standard versions please contact the support team.</p>
```
Comment (SawNawDfE) [inline src/SFA.DAS.AssessorService.Web/Views/Standard/_OptOutInsetPartial.cshtml:10]: Content changed as discussed
